lib/maths/CBoostedTreeFactory.cc

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CBoostedTreeFactory.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTreeImpl.h>
+
+namespace ml {
+namespace maths {
+
+CBoostedTreeFactory::TBoostedTreeUPtr CBoostedTreeFactory::build() {
+    this->initializeMissingFeatureMasks(*m_Frame);
+    std::tie(m_Tree->m_Impl->m_TrainingRowMasks,
+             m_Tree->m_Impl->m_TestingRowMasks) = this->crossValidationRowMasks();
+
+    // We store the gradient and curvature of the loss function and the predicted
+    // value for the dependent variable of the regression.
+    m_Frame->resizeColumns(m_Tree->m_Impl->m_NumberThreads, m_Frame->numberColumns() + 3);
+
+    this->initializeFeatureSampleDistribution(*m_Frame);
+    this->initializeHyperparameters(*m_Frame, m_ProgressCallback);
+    this->m_Tree->m_Impl->m_BayesianOptimization =
+        std::make_unique<CBayesianOptimisation>(this->hyperparameterBoundingBox());
+    this->m_Tree->m_Impl->m_NumberRounds = this->numberHyperparameterTuningRounds();
+
+    this->m_Tree->m_Impl->m_CurrentRound = 0; // for first start
+    return std::move(m_Tree);
+}
+
+CBoostedTreeFactory::operator TBoostedTreeUPtr() {
+    return std::move(this->build());
```

Comment:
```suggestion
    return this->build();
```
This prevents copy elision of temporary by forcing it to use move constructor.